src/wavm/openmp.cpp

```diff
@@ -378,188 +380,154 @@ WAVM_DEFINE_INTRINSIC_FUNCTION(env,
+    size_t snapshotSize = parentModule->snapshotToState(snapshotKey);
+    logger->debug(
+      "Created OpenMP snapshot: {} ({} bytes)", snapshotKey, snapshotSize);
 
-        const faabric::Message* originalCall = getExecutingCall();
-        const std::string origStr =
-          faabric::util::funcToString(*originalCall, false);
+    const faabric::Message* originalCall = getExecutingCall();
+    const std::string origStr =
+      faabric::util::funcToString(*originalCall, false);
 
-        U32* nativeArgs =
-          Runtime::memoryArrayPtr<U32>(memoryPtr, argsPtr, argc);
-
-        // Make the chained calls
-        auto& sch = faabric::scheduler::getScheduler();
-        for (int threadNum = 0; threadNum < nextNumThreads; threadNum++) {
-            faabric::Message call = faabric::util::messageFactory(
-              originalCall->user(), originalCall->function());
-
-            call.set_isasync(true);
-
-            for (int argIdx = argc - 1; argIdx >= 0; argIdx--) {
-                call.add_ompfunctionargs(nativeArgs[argIdx]);
-            }
-
-            // Snapshot details
-            call.set_snapshotkey(activeSnapshotKey);
-            call.set_snapshotsize(threadSnapshotSize);
-
-            call.set_funcptr(microtaskPtr);
-
-            call.set_ompthreadnum(threadNum);
-            call.set_ompnumthreads(nextNumThreads);
-
-            call.set_ompdepth(nextLevel->depth);
-            call.set_ompeffdepth(nextLevel->activeLevels);
-            call.set_ompmal(nextLevel->maxActiveLevels);
-
-            sch.callFunction(call);
-
-            const std::string chainedStr =
-              faabric::util::funcToString(call, false);
-            logger->debug("Forked thread {} ({}) -> {} {}(*{}) ({})",
-                          origStr,
-                          faabric::util::getSystemConfig().endpointHost,
-                          chainedStr,
-                          microtaskPtr,
-                          argsPtr,
-                          call.scheduledhost());
+    // Get pointers to shared variables in host memory
+    U32* sharedVarsPtr = nullptr;
+    if (argc > 0) {
+        sharedVarsPtr = Runtime::memoryArrayPtr<U32>(memoryPtr, argsPtr, argc);
+    }
 
-            chainedThreads[threadNum] = call.id();
+    // Set up the chained calls
+    std::vector<faabric::Message> msgs;
+    std::vector<int> callIds;
+    for (int threadNum = 0; threadNum < nextNumThreads; threadNum++) {
+        // Create basic call
+        faabric::Message call = faabric::util::messageFactory(
+          originalCall->user(), originalCall->function());
+
+        // All calls are async by definition
+        call.set_isasync(true);
+
+        // Snapshot details
+        call.set_snapshotkey(snapshotKey);
+        call.set_snapshotsize(snapshotSize);
+
+        // Function pointer
+        call.set_funcptr(microtaskPtr);
+
+        // Args
+        // TODO - is the order of these right? Does protobuf push to the back or
```

Comment:
Is this solved? Seems like a breaking choice.

Reply:
Yes good q, it seems to work for now but I will be digging into it more with the next phase of OpenMP work, so will leave it there as as note to self.